### ISSUE TITLE
Blocking annotation

### ISSUE THREAD

**Issue by tkaitchuck** [ceylon/ceylon-spec#1351]
In order to add future support for automatic thread safety and to be inform developers looking at an interface it is useful to denote which methods block. 

I've created a visitor to add support for this:
https://github.com/tkaitchuck/ceylon-spec/commit/e5da6c4d359f3a53ff1a2c464f8a371ac7bb86ad

Can someone look this over as this is my first contribution and I don't really fully understand things yet?

I assume I will need to define the annotation somewhere but I have not figured out where these things go.

Proposed definition below:
shared final annotation class BlockingAnnotation() satisfies OptionalAnnotation<BlockingAnnotation, FunctionOrValueDeclaration> {}
shared annotation BlockingAnnotation blocking() => BlockingAnnotation();

**Comment by tkaitchuck** [comment]
Also if anyone has feedback on how to improve this, please post.

Because the annotation is likely to be annoyingly common, it may be worthwhile excepting methods where it can be inferred. (IE: Private methods, run methods) Then it would only need to be done for shared methods.

**Comment by tombentley** [comment]
> I assume I will need to define the annotation somewhere but I have not figured out where these things go.

In `ceylon.language` there's a `annotations.ceylon` source file where all the language module annotations are defined.

> Because the annotation is likely to be annoyingly common, it may be worthwhile excepting methods where it can be inferred. (IE: Private methods, run methods) Then it would only need to be done for shared methods.

This could be a problem, if virtually every `shared` method has to become `shared blocking`. Why can't you infer it always?

**Comment by simonthum** [comment]
What's the point of this? Do you plan to add a `nonblocking` annotation to check for violators?

BTW I once made a proposal to annotate the sideeffectiness of functions and properties. But for autmatic thread safety, you'd need this and more ;)

**Comment by tkaitchuck** [comment]
@tombentley 
Yes. Thinking about it a bit more, I think its application can be totally automated. I'll still need the annotation for native methods that can't be intersected. But for all other methods instead of making it a compile error, I can simply have the compiler add the annotation. 

I'll get started on this approach.

**Comment by tkaitchuck** [comment]
I had assumed that compiler annotations would allow annotations to be added during compilation. Apparently this is not the case. Is there a good way to do that?

**Comment by gavinking** [comment]
@tkaitchuck We don't have any API to just attach untyped information to the AST. What we do instead is add attributes to the model objects in `ceylon-model`, or, in extreme cases, to the AST `Node` types.
